src/tools/perf/lib/uct_tests.cc

```diff
@@ -119,6 +119,37 @@ class uct_perf_test_runner {
         return length;
     }
 
+    inline void set_sn(void *dst_sn,
+                       ucs_memory_type_t dst_mem_type,
+                       const void *src_sn,
+                       ucs_memory_type_t src_mem_type) const {
+        if (ucs_likely(m_perf.allocator->mem_type == UCS_MEMORY_TYPE_HOST)) {
+            ucs_assert((dst_mem_type == UCS_MEMORY_TYPE_HOST) &&
+                       (src_mem_type == UCS_MEMORY_TYPE_HOST));
+            *(psn_t*)dst_sn = *(psn_t*)src_sn;
+        } else {
+            m_perf.allocator->memcpy(dst_sn, dst_mem_type,
+                                     src_sn, src_mem_type,
+                                     sizeof(psn_t));
+        }
+    }
+
+    inline psn_t get_sn(const void *sn) const {
+        if (ucs_likely(m_perf.allocator->mem_type == UCS_MEMORY_TYPE_HOST)) {
+            return *(psn_t*)sn;
+        } else {
+            psn_t host_sn;
+            m_perf.allocator->memcpy(&host_sn, UCS_MEMORY_TYPE_HOST,
+                                     sn, m_perf.allocator->mem_type,
+                                     sizeof(psn_t));
+            return host_sn;
+        }
+    }
+
+    inline psn_t get_sn(volatile psn_t *sn) const {
```

Comment:
why need to add volatile? seems this wrapper is not really needed

Reply:
this is needed to fix the following:
```
uct_tests.cc:149:18: note:   no known conversion for argument 1 from ‘volatile psn_t* {aka volatile unsigned char*}’ to ‘uct_perf_test_runner<(ucx_perf_cmd_t)3u, (ucx_perf_test_type_t)0u, (uct_perf_data_layout_t)0u, true>::psn_t* {aka unsigned char*}’
uct_tests.cc:537:40: error: call of overloaded ‘get_sn(volatile psn_t*&)’ is ambiguous
                     sn = get_sn(recv_sn);
                                        ^
uct_tests.cc:537:40: note: candidates are:
uct_tests.cc:137:18: note: uct_perf_test_runner<CMD, TYPE, DATA, ONESIDED>::psn_t uct_perf_test_runner<CMD, TYPE, DATA, ONESIDED>::get_sn(const void*) const [with ucx_perf_cmd_t CMD = (ucx_perf_cmd_t)3u; ucx_perf_test_type_t TYPE = (ucx_perf_test_type_t)0u; uct_perf_data_layout_t DATA = (uct_perf_data_layout_t)0u; bool ONESIDED = true; uct_perf_test_runner<CMD, TYPE, DATA, ONESIDED>::psn_t = unsigned char] <near match>
     inline psn_t get_sn(const void *sn) const {
```

`recv_sn` is declared as `volatile psn_t *recv_sn;`

Reply:
maybe change get_sn to accept const volatile void*?

Reply:
> maybe change get_sn to accept const volatile void*?

then we will need to do (i.e. `const_cast` before doing `reinterpret_cast`) even there when we don't need it (e.g. when `sn` is declared w/o `volatile`)
```
return *reinterpret_cast<const psn_t*>(const_cast<const void*>(sn));
```

Reply:
is it ok? or better have as is?

Reply:
how about
`return *reinterpret_cast<const volatile psn_t*>(sn);`

Reply:
> how about
> `return *reinterpret_cast<const volatile psn_t*>(sn);`

fixed

Reply:
so can we remove this wrapper function now?
also pls squash